File: services/ads-engine/otr_ads/schemas.py
from datetime import date
from typing import Literal
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field, model_validator

class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", allow_inf_nan=False)

class Metrics(StrictModel):
    source: Literal["demo", "manual", "meta", "google", "tiktok"]
    periodStart: date
    periodEnd: date
    spend: float = Field(ge=0)
    impressions: int = Field(ge=0)
    clicks: int = Field(ge=0)
    leads: int = Field(ge=0)
    customers: int = Field(ge=0)
    revenue: float = Field(ge=0)
    @model_validator(mode="after")
    def dates(self):
        if self.periodStart > self.periodEnd:
            raise ValueError("Invalid reporting period")
        return self

class Evidence(StrictModel):
    # User-attested observations. Omitted values are unknown, never false.
    trackingVerified: bool | None = None
    creativeFatigueObserved: bool | None = None
    budgetOnPace: bool | None = None
    audienceVerified: bool | None = None
    landingPageVerified: bool | None = None
    performanceOnTarget: bool | None = None
    note: str = Field(default="", max_length=1000)

class AuditInput(StrictModel):
    businessId: UUID
    campaignId: UUID | None = None
    metrics: list[Metrics] = Field(default_factory=list, max_length=100)
    evidence: Evidence = Field(default_factory=Evidence)

class Finding(StrictModel):
    control_id: str
    category: str
    status: Literal["pass", "fail", "unknown"]
    severity: Literal["high", "medium"]
    confidence: Literal["medium", "none"]
    observation: str
    evidence: list[dict]

class Recommendation(StrictModel):
    id: UUID
    title: str
    description: str
    category: str
    severity: Literal["high", "medium"]
    confidence: Literal["medium"]
    evidence: list[dict]
    businessId: UUID
    campaignId: UUID | None
    proposedAction: Literal["refresh_creative", "review_evidence"]
    requiresApproval: Literal[True] = True
    status: Literal["Suggested"] = "Suggested"

class AuditReport(StrictModel):
    schemaVersion: Literal[1] = 1
    status: Literal["normal", "provisional", "insufficient_evidence"]
    metricContext: Metrics | None = None
    healthScore: float | None
    evidenceCoverage: float
    findings: list[Finding]
    recommendations: list[Recommendation]
    dataGaps: list[str]
    sources: list[str]
    externalExecutionEnabled: Literal[False] = False

File: src/lib/ads/schemas.ts
import { z } from "zod";
import { performanceMetricSchema } from "../domain/schemas";
export const evidenceKeys = [
  "trackingVerified",
  "creativeFatigueObserved",
  "budgetOnPace",
  "audienceVerified",
  "landingPageVerified",
  "performanceOnTarget",
] as const;
export const evidenceLabels = [
  "Tracking verified",
  "Creative fatigue observed",
  "Budget on pace",
  "Audience verified",
  "Landing page verified",
  "Performance on target",
];
export const evidenceSchema = z
  .object({
    trackingVerified: z.boolean().nullable().optional(),
    creativeFatigueObserved: z.boolean().nullable().optional(),
    budgetOnPace: z.boolean().nullable().optional(),
    audienceVerified: z.boolean().nullable().optional(),
    landingPageVerified: z.boolean().nullable().optional(),
    performanceOnTarget: z.boolean().nullable().optional(),
    note: z.string().max(1000).default(""),
  })
  .strict();
const evidence = z.array(z.record(z.string(), z.unknown())).max(100);
export const recommendationSchema = z.object({
  id: z.uuid(),
  title: z.string().max(300),
  description: z.string().max(5000),
  category: z.string().max(100),
  severity: z.enum(["high", "medium"]),
  confidence: z.literal("medium"),
  evidence,
  businessId: z.uuid(),
  campaignId: z.uuid().nullable(),
  proposedAction: z.enum(["refresh_creative", "review_evidence"]),
  requiresApproval: z.literal(true),
  status: z.literal("Suggested"),
});
export type Recommendation = z.infer<typeof recommendationSchema>;
export const auditReportSchema = z.object({
  schemaVersion: z.literal(1),
  metricContext: performanceMetricSchema
    .omit({ id: true, businessId: true })
    .nullable()
    .default(null),
  status: z.enum(["normal", "provisional", "insufficient_evidence"]),
  healthScore: z.number().min(0).max(100).nullable(),
  evidenceCoverage: z.number().min(0).max(100),
  findings: z
    .array(
      z.object({
        control_id: z.string(),
        category: z.string(),
        status: z.enum(["pass", "fail", "unknown"]),
        severity: z.enum(["high", "medium"]),
        confidence: z.enum(["medium", "none"]),
        observation: z.string(),
        evidence,
      }),
    )
    .max(100),
  recommendations: z.array(recommendationSchema).max(100),
  dataGaps: z.array(z.string()),
  sources: z.array(z.string()),
  externalExecutionEnabled: z.literal(false),
});
export type AuditReport = z.infer<typeof auditReportSchema>;
